---
output:
  pdf_document: default
  html_document: default
---
#MetaScope Testing
##Installation & Including Required Libraries
```{r setup}

if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")
devtools::install_github("wejlab/MetaScope")

```

##Supressing Warning Messages and Including NCBI key for Ref data 
```{r lib}

suppressPackageStartupMessages({
  library(MetaScope)
  library(magrittr)
  library(Biostrings)
})
NCBI_key <- "01d22876be34df5c28f4aedc479a2674c809"
options("ENTREZ_KEY" = NCBI_key)

```

##Downloadiing Target Genomes
```{r tar_download}

target_ref_temp <- tempfile()
dir.create(target_ref_temp)

all_species <- c("Mycobacterium")
all_species <- c("Streptococcus")
#all_species <- c("Staphylococcus aureus RF122",
#                 "Staphylococcus aureus subsp. aureus ST398",
#                 "Staphylococcus aureus subsp. aureus Mu3")
#all_species <- c("Bacteria",
#                 "Fungi",
#                 "Viruses")
# sapply(all_species, download_refseq, 
#       reference = FALSE, representative = FALSE, compress = TRUE,
#       out_dir = target_ref_temp, caching = TRUE)

```

##Downloading Filter Genomes
```{r ref_download}

filter_ref_temp <- tempfile()
dir.create(filter_ref_temp)

#download_refseq(
#  taxon = "Staphylococcus epidermidis RP62A",
#  representative = FALSE, reference = FALSE,
#  compress = TRUE, out_dir = filter_ref_temp,
#  caching = TRUE)

```

##DeMultiplexing
```{r demult}

# Get barcode, index, and read data locations
barcodePath <-
  system.file("extdata", "barcodes.txt", package = "MetaScope")
indexPath <- system.file("extdata", "virus_example_index.fastq",
                         package = "MetaScope")
readPath <-
  system.file("extdata", "virus_example.fastq", package = "MetaScope")

# Get barcode, index, and read data locations
demult <-
  demultiplex(barcodePath,
              indexPath,
              readPath,
              rcBarcodes = FALSE,
              hammingDist = 2,
              location = tempfile())

demult

```

##Alignment with Reference Library
###For Target
```{r Index}

# ## Create temp directory to store the Bowtie2 indices
# index_temp <- tempfile()
# dir.create(index_temp)

# Create target index
mk_bowtie_index(
  ref_dir = "~/Documents/RStudio-git/TestingPhase/Ref_dir/",
  lib_dir = "~/Documents/RStudio-git/TestingPhase/Ref_dir/",
  lib_name = "target",
  overwrite = TRUE
)

```
###For Filter
```{r Filter}

# Create filter index
mk_bowtie_index(
  ref_dir = "~/Documents/RStudio-git/TestingPhase/Filter_dir/",
  lib_dir = "~/Documents/RStudio-git/TestingPhase/Filter_dir/",
  lib_name = "filter",
  overwrite = TRUE
)

```
### Store Bam File
```{r Bam}

# Create a temp directory to store output bam file
output_temp <- tempfile()
dir.create(output_temp)

# Get path to example reads
readPath <-
  system.file("extdata", "reads.fastq", package = "MetaScope")

# Align reads to the target genomes
target_map <-
  align_target_bowtie(
    read1 = readPath,
    lib_dir = "~/Documents/RStudio-git/TestingPhase/Ref_dir/",
    libs = "target",
    align_dir = output_temp,
    align_file = "bowtie_target",
    overwrite = TRUE
  )

```